File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-1.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Node temp1 = head;
        Node curr = temp1;
        
        while (temp1 != null) {
            Node dummy = new Node(temp1.val);
            Node next = temp1.next;
            dummy.next = next;
            temp1.next = dummy;
            temp1 = next;
        }

        Node curr1 = curr;
        
        while (curr != null) {
            if (curr.next != null) {
                if(curr.random != null){
                    curr.next.random = curr.random.next;
                }else{
                    curr.next.random = null;
                }
                
                curr = curr.next.next;
            }
        }
        
        Node ans;
        if(curr1 != null){
            ans = curr1.next;
        }else{
            ans = null;
        } 
         while (curr1 != null) {
            if (curr1.next != null) {
                Node next = curr1.next.next;
                if(next != null){
                    curr1.next.next = next.next;
                }else{
                    curr1.next.next = null;
                }
                curr1.next = next;
                curr1 = next;
            }
        }

        return ans;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        if(head == null){
            return head
        }

        var fastNode:ListNode? = null
        var slowNode:ListNode? = null

        for(i: Int in 1 .. n){
            if(fastNode != null){
                fastNode = fastNode?.next
            }else{
                fastNode = head
            }
        }

        while(fastNode?.next != null){
            if(slowNode == null){
                slowNode = head
            }else{
                slowNode = slowNode?.next
            }
            fastNode = fastNode?.next
        }

        if(slowNode == null){
            return head?.next
        }else{
            slowNode?.next = slowNode?.next?.next
        }

        return head
    }
}
